pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -524,21 +532,31 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				set.Namespace,
 				set.Name,
 				replicas[target].Name)
-			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[target]); err != nil {
+				return nil, err
+			}
+			// After deleting a Pod, dont return from here yet.
+			// We might have maxUnavailable greater than 1.
 			status.CurrentReplicas--
-			return &status, err
 		}
 
-		// wait for unhealthy Pods on update
-		if !isHealthy(replicas[target]) {
+		if getPodRevision(replicas[target]) != updateRevision.Name || !isHealthy(replicas[target]) {
+			unavailablePods = append(unavailablePods, replicas[target].Name)
+
+		}
+
+		// If at anytime, total number of unavailable Pods exceeds maxUnavailable,
+		// we stop deleting more Pods for Update
+		if len(unavailablePods) >= maxUnavailable {
```

Comment:
I think you need to count all the pods before stopping any of them. The very first shutdown could have violated this, no? Do I misunderstand the feature?

Reply:
replied here https://github.com/kubernetes/kubernetes/pull/82162/files/45ec02a384b41abb86aaecd44fabb66e0e0f688c#r558754508

Reply:
this is resolved now

Reply:
This is not resolved by my reading of the code. Note: it's a different problem than the other one.

Please add a test that proves the order in which the pods are examined doesn't have an impact on whether or not pods are killed.

Reply:
You have to consider the possibility that maxUnavailable was changed by the user in the middle of the upgrade.

For example, let's say maxUnavailable=3, the controller deletes 3 pods: (n, n-1, n-2). Then n becomes ready. At this point (n-1, n-2) are still updating. Suddenly, the user realizes that there was an error in the image upgrade, so they upload a new one and set maxUnavailable=1 to be more cautious this time.

In this case, since n is available, your code will remove n, even though (n-1, n-2) are still down.

That's why you should count the unavailable pods first.

Reply:
So, due to the loop above (before the changes in this PR), isHealthy should always return false. Why are we checking again then? Just in case?